Functions agent

danny-avila opened the pull request:
Experimental, not fully optimized, but this will be the de facto manner once the optimizations and improvements are made.

## **New Feature as of today 6/13: [Function calling and other API updates](https://openai.com/blog/function-calling-and-other-api-updates)**

![image](https://github.com/danny-avila/LibreChat/assets/110412045/891bc312-5d00-4369-869a-cf83a585a593)

![image](https://github.com/danny-avila/LibreChat/assets/110412045/75fecaf2-ec04-41a6-a781-302c702d487d)

**The immediate benefits:**
- faster tool selection
- less priming/token usage for selection,
- has room to be creative, so you can use higher temps on the Agent Settings

**Long-term benefit as we iterate on this:** 
- I can optimize the chain as well as the final output
- all existing/future tools will have massive improvements/reliability with multi-input calls
- I added a multi-input or **[StructuredTool](http://web.archive.org/web/20230518082852/https://blog.langchain.dev/structured-tools/)** version of **Stable Diffusion** so that the new functions-compatible model knows it accepts multiple inputs. Read more about it in the [langchain docs](https://python.langchain.com/en/latest/modules/agents/tools/multi_input_tool.html?highlight=structuredtool), note: this doesn't have the best documentation at the moment.
- As already mentioned, even with single input plugins, the new models overall better at tooling
- Todo: update docs for making a StructuredTool compatible with this project, and to reflect other changes made in this update
    - other changes are minor: tools directory was re-organized to be more coherent
- Todo: update all current plugins to multi-input versions

## Comparison Example 

**User prompt:** `Generate me a picture of Sunny` (a corgi discussed about in prior messages)
### Classic mode 
```bash
{
	plugin: Stable Diffusion
	input: photo of a Pembroke Welsh Corgi with a bright red coat, white markings on face, chest, and paws, perky ears, warm
	brown eyes, bushy tail, energetic, playful, social, intelligent, easy to train, loves cuddling, 8k UHD, natural lighting, happy,
	friendly | cartoon, sketch, low quality, ugly, mutation, deformed
	thought: The user wants a picture of Sunny, the corgi I just described. I can use the stable-diffusion tool to generate
	an image that matches the description.
}
```

**Pros:** Great at planning and multi-modality
**Cons:** High token use, limited to single inputs, unreliable response formats

### Functions mode
```bash
{
	plugin: Stable Diffusion
	input: {
	    "prompt":"photo of a Pembroke Welsh Corgi, red coat, white markings, perky ears, brown eyes, bushy tail, 
	    happy expression",
	    "negative_prompt":"cartoon, sketch, unrealistic, CGI"
	}
}
```

**Pros:** Highly Efficient, multi-input tooling, less token use, reliable response formats
**Cons:** unreliable with multi-modal queries, ends chains prematurely

**Conceptual To-Do:** Implement a solution that utilizes the strengths of both

I've improved the cons of the Functions Agent by using strong language in the instructions, but may still experience issues.

### What's changed

- A new Agent mode was created, in very similar format to OpenAIAgent from langchainjs with chat memory. The OpenAI agent was introduced to langchain with its introduction of the functions-compatible OpenAI models in this update: https://github.com/hwchase17/langchainjs/pull/1631
    - The code can be found `api\app\langchain\agents\Functions`
- Wired up the ChatAgent for handling of the new Functions agent mode
    - might be anecdotal but updated/new models are more obedient when referred to as Assistants, rather than 'ChatGPT', so all mentions of the latter in priming were removed.
    - intermediateSteps are handled differently due to object inputs for actions
    - loadTools is passed a boolean to determine if agent is in Functions mode
    - loadTools will switch to a StructuredTool version of an existing Tool if in functions mode
 - Added new test for loadTools testing the function above
 - Refactored/reorganized tools directory
 - Frontend/Backend were updated to handle the new agentOptions prop: `agent`

![image](https://github.com/danny-avila/LibreChat/assets/110412045/fb438111-3c64-4491-8ee0-659f4900d573)

danny-avila committed:
feat(experimental): FunctionsAgent, uses new function payload for tooling

danny-avila committed:
feat(frontend): add support for agent selection in GPT plugins in adding functions agent

Add support for selecting an agent in the GPT plugins endpoint. The agent can be selected from a dropdown menu in the AgentSettings component. The default agent is set to 'classic' in the cleanupPreset, getDefaultConversation, and handleSubmit functions.

danny-avila committed:
refactor(FunctionsAgent.js): change var to const in plan function

danny-avila committed:
refactor(ChatAgent.js, handlers.js): improve logging format and add support for functionsAgent

- Improve logging format in ChatAgent.js by adding more details to the log
- Add support for functionsAgent in ChatAgent.js to format the log differently
- Improve formatAction function in handlers.js to handle empty thoughts and add support for functionsAgent

danny-avila committed:
refactor(tools): restructure tool dir

danny-avila committed:
feat(StableDiffusion.js): add StableDiffusionAPI as a StructuredTool for Functions Agent

danny-avila committed:
refactor(langchain/tools): move availableTools import to tools/index.js

danny-avila committed:
refactor(handleTools.js): change loadTools function signature to include functions parameter
feat(handleTools.test.js): add test for loading StructuredSD tool with functions parameter

danny-avila committed:
refactor(StableDiffusion.js): update prompt and negative_prompt schema descriptions to include minimum number of keywords
fix(StableDiffusion.js): fix output path for generated image
feat(askGPTPlugins.js): update import path for validateTools function

danny-avila committed:
refactor(ChatAgent.js, handlers.js): stringify toolInput object in logs

The `toolInput` object was not being properly logged in the `ChatAgent.js` and `handlers.js` files. The `JSON.stringify()` method was added to properly log the object.

danny-avila committed:
fix: remove public/images directory and image created during testing

danny-avila committed:
feat(FunctionsAgent): improve LLM instructions for more reliable results

fuegovic commented:
@danny-avila your PR description is awesome

danorlando commented on api/app/langchain/tools/util/handleTools.test.js:
```diff
@@ -11,21 +11,20 @@ var mockPluginService = {
 });
 
-jest.mock('../../../server/services/PluginService', () => mockPluginService);
+jest.mock('../../../../server/services/PluginService', () => mockPluginService);
```
Just a thought, but it would be nice if we could have aliases like we do for the api to avoid getting stuck in dot-dot-slash hell. Obviously there is no bundler but there has to be a better way to navigate the folder structure.

danorlando replied:
I take that back. There's other ways of doing it, but none that are as convenient as just setting aliases unforuntately.

danorlando commented:
> @danny-avila your PR description is awesome

Yea, I have to agree. Thanks for doing that. I'm constantly amazed at how you're able to stay at the forefront of emerging technologies because I hadn't even heard of Function agents yet and you already had a PR up for it, so this description was really helpful for understanding what this new feature was.

ClaraLeigh reviewed (approved):
Looks good! my only gripe would be the commented out code, if we aren't using it anymore it would be better to keep things clean

fuegovic reviewed (approved):
From a user’s perspective, everything is good. I have not encountered any issues from testing it.

danny-avila committed:
refactor(FunctionsAgent.js): remove unnecessary text from PREFIX constant

danny-avila committed:
feat(ChatAgent.js): add support for skipping completion mode in ChatAgent
feat(ChatAgent.js): add a check for images when completion is skipped to add to response
feat(askGPTPlugins.js): add skipCompletion option to agentOptions
feat(client): add Switch component to ui components and use for new Agent Settings
chore(package.json): ignore client directory in nodemonConfig

danny-avila committed:
feat(tools): add structured Wolfram tool

danny-avila committed:
style(OptionHover.jsx): add descriptions for new types
feat(AgentSettings.jsx): change OptionHover type from 'temp' to 'func' and from 'temp' to 'skip'
